python/lsst/daf/butler/core/_butlerUri.py

```diff
@@ -0,0 +1,1332 @@
+        return (str(self),)
+
+    @staticmethod
+    def _fixupPathUri(parsed: urllib.parse.ParseResult, root: Optional[str] = None,
+                      forceAbsolute: bool = False,
+                      forceDirectory: bool = False) -> Tuple[urllib.parse.ParseResult, bool]:
+        """Correct any issues with the supplied URI.
+
+        Parameters
+        ----------
+        parsed : `~urllib.parse.ParseResult`
+            The result from parsing a URI using `urllib.parse`.
+        root : `str`, ignored
+            Not used by the this implementation since all URIs are
+            absolute except for those representing the local file system.
+        forceAbsolute : `bool`, ignored.
+            Not used by this implementation. URIs are generally always
+            absolute.
+        forceDirectory : `bool`, optional
+            If `True` forces the URI to end with a separator, otherwise given
+            URI is interpreted as is. Specifying that the URI is conceptually
+            equivalent to a directory can break some ambiguities when
+            interpreting the last element of a path.
+
+        Returns
+        -------
+        modified : `~urllib.parse.ParseResult`
+            Update result if a URI is being handled.
+        dirLike : `bool`
+            `True` if given parsed URI has a trailing separator or
+            forceDirectory is True. Otherwise `False`.
+
+        Notes
+        -----
+        Relative paths are explicitly not supported by RFC8089 but `urllib`
+        does accept URIs of the form ``file:relative/path.ext``. They need
+        to be turned into absolute paths before they can be used.  This is
+        always done regardless of the ``forceAbsolute`` parameter.
+
+        AWS S3 differentiates between keys with trailing POSIX separators (i.e
+        `/dir` and `/dir/`) whereas POSIX does not neccessarily.
+
+        Scheme-less paths are normalized.
+        """
+        # assume we are not dealing with a directory like URI
+        dirLike = False
+
+        # URI is dir-like if explicitly stated or if it ends on a separator
+        endsOnSep = parsed.path.endswith(posixpath.sep)
+        if forceDirectory or endsOnSep:
+            dirLike = True
+            # only add the separator if it's not already there
+            if not endsOnSep:
+                parsed = parsed._replace(path=parsed.path+posixpath.sep)
+
+        return parsed, dirLike
+
+    def transfer_from(self, src: ButlerURI, transfer: str,
+                      transaction: Optional[Union[DatastoreTransaction, NoTransaction]] = None) -> None:
+        """Transfer the current resource to a new location.
+
+        Parameters
+        ----------
+        src : `ButlerURI`
+            Source URI.
+        transfer : `str`
+            Mode to use for transferring the resource. Generically there are
+            many standard options: copy, link, symlink, hardlink, relsymlink.
+            Not all URIs support all modes.
```

Comment:
Shouldn't this be an enum?

Reply:
Probably true. I think I'll punt to a new ticket because there are ramifications for the APIs that pass the transfer mode into this routine that I need to consider.